### PR TITLE
represents instead of presenter

### LINKED ISSUE #61: allow attribute to return array when dependency is empty
example:

``` ruby
    attribute :zones do
      return [] unless row_model.zone_tag # need an option for dependencies to return [] or nil in csv_row_model
      context.project.zones.tagged_with(row_model.zone_tag)
    end
```

### PR DESCRIPTION
**Pull request by s12chung**
see README

closes #61 

just need to do a final cleanup here: https://github.com/FinalCAD/csv_row_model/blob/c7b73c9bdf7a50d771de951f299155daaca6c5fb/lib/csv_row_model/import/base.rb#L116-L119

will update our code when people aren't writing build code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -279,61 +240,84 @@ row_model.errors.full_messages # => ["Id must be greater than 0"]
...
 1. Memoizes by default, turn off with `memoize: false` option
-2. All errors of `row_model` are propagated to the presenter when calling `presenter.valid?`
-3. Handles dependencies:
-  - When any of the dependencies are `blank?`, the attribute block is not called and the attribute returns `nil`.
-  - When any of the dependencies are `invalid?`, `presenter.errors` for dependencies are cleaned. For the example above, if `row_model.id/name` are `invalid?`, then
-the `:project` key is removed from the errors, so: `presenter.errors.keys # => [:id, :name]`
+2. Handles dependencies:
+  - When any of the dependencies are `blank?`, the attribute block is not called and the representation returns `nil`.
+  - When any of the dependencies are `invalid?`, `row_model.errors` for dependencies are cleaned. For the example above, if `id/name` are `invalid?`, then
+the `:project` key is removed from the errors, so: `row_model.errors.keys # => [:id, :name]` (applies to warnings as well)
```
Comment (joel) [inline README.md:294]: 👍

File: lib/csv_row_model/import/represents.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      protected
+
+      # remove each dependent attribute from errors if it's representation dependencies are in the errors
+      def filter_errors
+        self.class.representation_names.each do |representation_name|
+          next unless errors.messages[representation_name] &&
+            errors.messages.slice(*self.class.representations[representation_name][:dependencies]).present?
+          errors.delete representation_name
+        end
+      end
```
Comment (joel) [inline lib/csv_row_model/import/represents.rb:33]: 👍

File: spec/csv_row_model/import/represents_spec.rb [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      it "calls #filter_errors" do
+        expect(instance).to receive(:filter_errors)
+        subject
+      end
+    end
+
+    describe "#filter_errors" do
+      subject { instance.send :filter_errors }
+      before do
+        klass.send(:represents_one, :test_model, dependencies: %i[string1]) { "test" }
```
Comment (joel) [inline spec/csv_row_model/import/represents_spec.rb:50]: `klass.send(:represents_one, :test_model, dependencies: :a_dependent_attribute) { 'test' }`

File: spec/csv_row_model/import/represents_spec.rb [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        instance.errors.add(:test_model)
+        instance.errors.add(:string2)
```
Comment (joel) [inline spec/csv_row_model/import/represents_spec.rb:52]: `instance.errors.add(:whatever_attribute)`

File: spec/csv_row_model/import/represents_spec.rb [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+          before do
+            instance.warnings.add(:test_model)
+            instance.warnings.add(:string2)
+            instance.warnings.add(:string1)
+          end
+
+          it "removes warnings only" do
+            subject
+            expect(instance.errors.keys).to eql [:test_model, :string2, :string1]
+            expect(instance.warnings.keys).to eql [:string2, :string1]
```
Comment (joel) [inline spec/csv_row_model/import/represents_spec.rb:80]: I'm not sure to understand why `instance.errors.keys` keep `:test_model` representation error in this context....

File: spec/csv_row_model/import/represents_spec.rb [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+          end
+        end
+      end
+    end
+
+    describe "#memoize" do
+      subject { instance.send(:memoize, "test") { Random.rand } }
+
+      it "memoizes the result" do
+        expect(subject).to_not eql nil
```
Comment (joel) [inline spec/csv_row_model/import/represents_spec.rb]: I guess this line could be removed

Comment (joel) [inline spec/csv_row_model/import/represents_spec.rb]: `expect(subject).to be_present`

**Comment by joel** [comment]
Very nice work, clear spec, easy to read, thanks

File: spec/csv_row_model/import/represents_spec.rb (hunk 5 above)
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb:80]: trying to test if `#safe?` accidentally touches errors

```
context "when called within #using_warnings" do
   subject { instance.using_warnings { super() } }
```

is equal to: https://github.com/s12chung/active_warnings/blob/master/lib/active_warnings.rb#L55-L58

File: spec/csv_row_model/import/represents_spec.rb (hunk 3 above)
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb:50]: `string1` is a column in `BasicImportModel` (`klass` subclasses this), yours is easier to read. i'm not sure if it's worth making a column just for the test :p

File: spec/csv_row_model/import/represents_spec.rb (hunk 6 above)
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb]: will change. yeah, easier to read.

File: spec/csv_row_model/import/represents_spec.rb (hunk 5 above)
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb:80]: i'll just call `#safe` as the subject instead. it's easier :p

File: spec/csv_row_model/import/represents_spec.rb [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    end
+
+    describe "#valid?" do
+      subject { instance.valid? }
+
+      it "calls #filter_errors" do
+        expect(instance).to receive(:filter_errors)
+        subject
+      end
+    end
```
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb:45]: hmm... i'll make this test a bit better.

File: spec/csv_row_model/import/represents_spec.rb [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+      context "with memoization off" do
+        let(:options) { { memoize: false } }
+
+        it "doesn't memoize anything" do
+          subject
+          expect(instance.test_model).to eql "test"
+          expect(instance.test_model.object_id).to_not eql instance.test_model.object_id
+          expect(instance.__test_model).to eql "test"
+          expect(instance.__test_model).to_not eql instance.__test_model.object_id
```
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb:214]: note to self. whops.

File: spec/csv_row_model/import/represents_spec.rb (hunk 5 above)
Comment (s12chung) [inline spec/csv_row_model/import/represents_spec.rb:80]: ah, can't be done with these tests because i want to isolate the method and filter the errors.

**Pull request merged**